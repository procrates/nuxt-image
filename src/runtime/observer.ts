export function createObserver (intersectionOptions: object) {
  const callbackType = { intersect: 'onIntersect', print: 'onPrint' }
  const elementCallbackMap = {}
  function intersectionCallback (entries, imgObserver) {
    entries.forEach((entry) => {
      if (entry.isIntersecting) {
        const lazyImage = entry.target
        const callback = elementCallbackMap[lazyImage.__unique]
        if (typeof callback === 'function') {
          callback(callbackType.intersect)
        }
        delete elementCallbackMap[lazyImage.__unique]
        imgObserver.unobserve(lazyImage)
      }
    })
  }

  printObserver(() => {
    Object.values(elementCallbackMap).forEach((callback: any) => callback(callbackType.print))
  })

  const intersectObserver = intersectionObserver(intersectionCallback, intersectionOptions)

  return {
    add (target, component, unique) {
      // add unique id to recognize target
      target.__unique = unique || target.id || target.__vue__._uid
      elementCallbackMap[target.__unique] = component
      intersectObserver.observe(target)
    },
    remove (target) {
      delete elementCallbackMap[target.__unique]
      intersectObserver.unobserve(target)
    }
  }
}

function intersectionObserver (onMatch, options): IntersectionObserver | undefined {
  if (typeof IntersectionObserver === 'undefined') {
    return
  }
  return new IntersectionObserver(onMatch, {
    rootMargin: '50px',
    ...options
  })
}

function printObserver (onMatch) {
  if (typeof window === 'undefined' || typeof window.matchMedia === 'undefined') {
    return
  }

  const mediaQueryList = window.matchMedia('print')
  mediaQueryList.addListener((query) => {
    if (query.matches) {
      onMatch()
    }
  })
}
