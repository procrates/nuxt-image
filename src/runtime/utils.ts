import { OperationGeneratorConfig } from 'types'

export function cleanDoubleSlashes (path) {
  return path.replace(/(https?:\/\/)|(\/)+/g, '$1$2')
}

export function createMapper (map: any) {
  return (key: string) => {
    return map[key] || key || map.missingValue
  }
}

export function createOperationsGenerator ({ formatter, keyMap, joinWith = '/', valueMap }: OperationGeneratorConfig = {}) {
  if (!formatter) {
    formatter = (key, value) => `${key}=${value}`
  }
  if (keyMap && typeof keyMap !== 'function') {
    keyMap = createMapper(keyMap)
  }
  valueMap = valueMap || {}
  Object.keys(valueMap).forEach((valueKey) => {
    if (typeof valueMap[valueKey] !== 'function') {
      valueMap[valueKey] = createMapper(valueMap[valueKey])
    }
  })

  return (modifiers: { [key: string]: string }) => {
    const operations = Object.entries(modifiers)
      .filter(([_, value]) => typeof value !== 'undefined')
      .map(([key, value]) => {
        const mapper = valueMap[key]
        if (typeof mapper === 'function') {
          value = mapper(modifiers[key])
        }

        key = typeof keyMap === 'function' ? keyMap(key) : key

        return formatter(key, value)
      })

    return operations.join(joinWith)
  }
}

type Attrs = { [key: string]: string|number }

export function renderAttributesToString (attributes: Attrs = {}) {
  return Object.entries(attributes)
    .map(([key, value]) => value ? `${key}="${value}"` : '')
    .filter(Boolean).join(' ')
}

export function renderTag (tag: string, attrs: Attrs, contents?: string) {
  const html = `<${tag} ${renderAttributesToString(attrs)}>`
  if (!contents) {
    return html
  }
  return html + contents + `</${tag}>`
}

const modernFormats = ['webp', 'avif']

export function isModernFormat (path: string) {
  return modernFormats.includes(String(path).split('.').pop())
}
