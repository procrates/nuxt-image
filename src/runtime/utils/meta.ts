import type { ImageInfo, ImageCTX } from '../../types/image'

export async function imageMeta (ctx: ImageCTX, url: string): Promise<ImageInfo> {
  const cache = getCache<ImageInfo>(ctx)

  const cacheKey = 'image:meta:' + url
  if (cache.has(cacheKey)) {
    return cache.get(cacheKey)
  }

  const meta = await _imageMeta(url).catch((err) => {
    // eslint-disable-next-line no-console
    console.error('Failed to get image meta for ' + url, err + '')
    return {
      width: 0,
      height: 0,
      ratio: 0
    }
  })

  cache.set(cacheKey, meta)
  return meta
}

async function _imageMeta (url: string): Promise<ImageInfo> {
  if (process.server) {
    const imageMeta = await import('image-meta').then(r => r.default || r)
    const data: Buffer = await fetch(url).then((res: any) => res.buffer())
    const metadata = imageMeta(data)
    if (!metadata) {
      throw new Error(`No metadata could be extracted from the image \`${url}\`.`)
    }
    const { width, height } = metadata
    const meta = {
      width: width!,
      height: height!,
      ratio: width && height ? width / height : undefined
    }

    return meta
  }
  if (typeof Image === 'undefined') {
    throw new TypeError('Image not supported')
  }

  return new Promise((resolve, reject) => {
    const img = new Image()
    img.onload = () => {
      const meta = {
        width: img.width,
        height: img.height,
        ratio: img.width / img.height
      }
      resolve(meta)
    }
    img.onerror = err => reject(err)
    img.src = url
  })
}

interface Cache<T> {
  get: (id: string) => T,
  set: (id: string, value: T) => void,
  has: (id: string) => boolean
}

function getCache <T = any> (ctx: ImageCTX): Cache<T> {
  if (!ctx.nuxtContext.cache) {
    if (ctx.nuxtContext.ssrContext && ctx.nuxtContext.ssrContext.cache) {
      ctx.nuxtContext.cache = ctx.nuxtContext.ssrContext.cache
    } else {
      const _cache: Record<string, any> = {}
      ctx.nuxtContext.cache = {
        get: (id: string) => _cache[id],
        set: (id: string, value: any) => { _cache[id] = value },
        has: (id: string) => typeof _cache[id] !== 'undefined'
      }
    }
  }
  return ctx.nuxtContext.cache
}
