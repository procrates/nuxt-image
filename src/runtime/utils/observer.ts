/* eslint-disable no-use-before-define */
type ObserverCallback = (event?: 'unsupported' | 'intersect') => any

interface Observer {
  add: (el: ObservedElement, cb: ObserverCallback) => void
  remove: (el: ObservedElement) => void
  supported: boolean
}

let observerInstance: Observer
let observerIdCtr = 1
const OBSERVER_ID_KEY = '__observer_id__'

type ObservedElement = Element & { [OBSERVER_ID_KEY]?: number | string }

export function getObserver (): Observer | false {
  if (typeof IntersectionObserver === 'undefined') {
    return false
  }

  if (observerInstance) {
    return observerInstance
  }

  const elementCallbackMap: Record<string | number, ObserverCallback> = {}

  const remove = (el: ObservedElement) => {
    const key = el[OBSERVER_ID_KEY]
    if (!key) { return }
    delete elementCallbackMap[key]
    delete el[OBSERVER_ID_KEY]
    intersectObserver.unobserve(el)
  }

  const add = (el: ObservedElement, fn: () => any) => {
    el[OBSERVER_ID_KEY] = el[OBSERVER_ID_KEY] || ++observerIdCtr
    elementCallbackMap[el[OBSERVER_ID_KEY]!] = fn
    intersectObserver.observe(el)
  }

  onPrint(() => {
    Object.values(elementCallbackMap).forEach((fn: any) => fn('print'))
  })

  const onMatch = (entries: IntersectionObserverEntry[]) => {
    for (const entry of entries) {
      if (entry.isIntersecting) {
        const fn = elementCallbackMap[(entry.target as ObservedElement)[OBSERVER_ID_KEY]!]
        if (typeof fn === 'function') { fn('intersect') }
        remove(entry.target)
      }
    }
  }
  const intersectObserver = new IntersectionObserver(onMatch, { rootMargin: '50px' })

  observerInstance = { supported: true, add, remove }
  return observerInstance
}

export function useObserver (el: ObservedElement, fn: ObserverCallback): Function {
  const observer = getObserver()
  if (!observer) {
    fn('unsupported')
    return () => {}
  }
  observer.add(el, fn)
  return () => observer.remove(el)
}

function onPrint (fn: ObserverCallback) {
  if (typeof window === 'undefined' || typeof window.matchMedia === 'undefined') {
    return
  }
  const mediaQueryList = window.matchMedia('print')
  mediaQueryList.addEventListener('change', (query) => {
    if (query.matches) {
      fn()
    }
  })
}
