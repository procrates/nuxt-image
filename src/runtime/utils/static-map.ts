const staticImageMap = {}

function updateImageMap () {
  if (typeof window.$nuxt !== 'undefined') {
    const pageImages = (window.$nuxt as any)._pagePayload?.data?.[0]?._img || {}
    Object.assign(staticImageMap, pageImages)
    console.log(staticImageMap)
  }
}

export function useStaticImageMap (nuxtContext?) {
  // Update on initialization
  updateImageMap()

  // Merge new mappings on route change
  if (nuxtContext) {
    nuxtContext.app.router.afterEach(updateImageMap)
  }

  // Make sure manifest is initialized
  if ((window as any).onNuxtReady) {
    (window as any).onNuxtReady(updateImageMap)
  }

  return staticImageMap
}
