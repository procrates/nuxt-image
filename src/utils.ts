import { resolve } from 'path'
import consola from 'consola'
import hasha from 'hasha'
import { name, version } from '../package.json'

export const logger = consola.withScope('@nuxt/image')

export const pkg = { name, version }

export const pkgDir = resolve(__dirname, '..')

export function hash (value: string, length = 6) {
  return hasha(value).substr(0, length)
}

// TODO: Typechecks
export function pick (obj: any, keys: string[]) {
  const newobj = {}
  for (const key of keys) {
    newobj[key] = obj[key]
  }
  return newobj
}
