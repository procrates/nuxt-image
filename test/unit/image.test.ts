/**
 * @jest-environment jsdom
 */

import type Vue from 'vue'
import { Wrapper } from '@vue/test-utils'

import { getSrc, mountWithImg } from './utils/mount'
import { mockObserver } from './utils/observer'
import { nextTick } from './utils/tick'

import NuxtImg from '~/runtime/components/nuxt-img.vue'

describe('Renders simple image', () => {
  let wrapper: Wrapper<Vue>
  const src = '/image.png'

  beforeEach(() => {
    wrapper = mountWithImg(NuxtImg, {
      width: 200,
      height: 200,
      sizes: '200,500:500,900:900',
      src
    })
  })

  test('Matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test.todo('alt attribute is generated')
  // () => {
  //   expect((wrapper.vm as any).generatedAlt).toEqual('image')
  //   const domAlt = wrapper.element.getAttribute('alt')
  //   expect(domAlt).toEqual('image')
  // }

  test('props.src is picked up by getImage()', () => {
    const domSrc = wrapper.element.getAttribute('src')
    expect(domSrc).toEqual(getSrc(src))
  })

  test('props.src is reactive', (done) => {
    const newSource = '/image.jpeg'
    wrapper.setProps({ src: newSource })
    process.nextTick(() => {
      const domSrc = wrapper.find('img').element.getAttribute('src')
      expect(domSrc).toEqual(getSrc(newSource))
      return done()
    })
  })

  test('sizes', () => {
    const sizes = wrapper.find('img').element.getAttribute('sizes')
    expect(sizes).toBe('(max-width: 500px) 500px, 900px')
  })

  test('registers an observer for lazy loading', async () => {
    const observer = mockObserver()
    wrapper = mountWithImg(NuxtImg, {
      loading: 'lazy',
      width: 200,
      height: 200,
      sizes: '200,500:500,900:900',
      src
    })

    expect(observer.wasAdded).toBe(true)
    expect(observer.wasDestroyed).toBe(false)

    expect(wrapper.html()).toMatchSnapshot()

    observer.triggerVisibility()

    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.destroy()
    expect(observer.wasDestroyed).toBe(true)
  })
})
