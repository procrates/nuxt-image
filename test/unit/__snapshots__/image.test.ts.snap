// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders simple image Matches snapshot 1`] = `"<img src=\\"/_custom/image.png\\" width=\\"200\\" height=\\"200\\" sizes=\\"(max-width: 500px) 500px, 900px\\" srcset=\\"/_custom/image.png 500w, /_custom/image.png 900w\\">"`;
